sympy/core/mod.py

```diff
@@ -36,12 +36,11 @@ def doit(p, q):
             to be less than or equal q.
             """
 
+            if q == S.Zero:
+                raise ZeroDivisionError("Modulo by zero")
             if p.is_infinite or q.is_infinite or p is nan or q is nan:
                 return nan
-            if (p == q or p == -q or
-                    p.is_Pow and p.exp.is_integer and p.base == q and q.is_integer
-                    and p.exp.is_positive or
-                    p.is_integer and q == 1):
+            if p == S.Zero or p == q or p == -q or (p.is_integer and q == 1):
```

Comment:
Is it intentional to use `p == S.Zero` which is true for p in (0, 0.0) and `q == 1` which is true for q in (1, 1.0)?

Reply:
@smichr, I just left original `q == 1` comparison and added `p == S.Zero` comparison in similar manner, but I don't think there is a problem.